File: sorting/python/shell.py
#!/usr/local/bin/python2
# -*- coding: UTF-8 -*-

def shell_sort(arr):
    n = len(arr)
    gap = n / 2
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]
            index = i
            for j in range(i - gap, -1, -gap):
                if arr[j] > temp:
                    arr[j + gap] = arr[j]
                    index = j
                else:
                    break
            arr[index] = temp
        gap /= 2
    return arr

def shell_sort2(arr):
    n = len(arr)
    gap = n / 2
    while gap > 0:
        for i in range(gap, n):
            temp = arr[i]
            j = i
            while (j - gap) >= 0 and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap
            arr[j] = temp
        gap /= 2
    return arr

if __name__ == "__main__":
    arr = [1, 3, 9, 0, 2, 4, 7, 8, 6, 5]

    print arr, "---default"
    # print shell_sort(arr), "---shell"
    print shell_sort2(arr), "---shell2"

File: sorting/golang/shell.go
package main

import (
	"fmt"
)

func shell_sort(s []int) []int {
	n := len(s)
	gap := n / 2
	for gap > 0 {
		for i := gap; i < n; i++ {
			temp := s[i]
			index := i
			for index >= gap && s[index-gap] > temp {
				s[index] = s[index-gap]
				index = index - gap
			}
			s[index] = temp
		}
		gap = gap / 2
	}
	return s
}

func main() {
	s := []int{1, 3, 9, 0, 2, 4, 7}
	fmt.Println(s, "---default")
	fmt.Println(shell_sort(s), "---shell")
}
